odoo/addons/base/models/ir_model.py

```diff
@@ -151,12 +140,23 @@ def jsonify(val):
         return Json({'en_US': val}) if val is not None else val
 
     wrappers = [(jsonify if model._fields[fname].translate else identity) for fname in fnames]
-    params = [
+    values = [
         tuple(func(val) for func, val in zip(wrappers, row))
         for row in rows
     ]
-    model.env.cr.execute(query, params)
-    return [row[0] for row in model.env.cr.fetchall()]
+    comma = SQL(", ").join
+    query = SQL("""
+        INSERT INTO %(table)s (%(cols)s) VALUES %(values)s
+        ON CONFLICT (%(conflict)s) DO UPDATE SET (%(cols)s) = (%(excluded)s)
+        RETURNING id
+        """,
+        table=SQL.identifier(model._table),
+        cols=comma(SQL.identifier(fname) for fname in fnames),
+        values=comma(SQL("%s", value) for value in values),
```

Comment:
`SQL.join()` can actually take parameters
```suggestion
        values=comma(values),
```